Sources/Algorithms/LazySplit.swift

```diff
@@ -0,0 +1,751 @@
+    var end = base.endIndex
+
+    if splitCount < maxSplits {
+      // The non-inclusive end of this subsequence is marked by the next
+      // separator, or the end of the base collection.
+      end =
+        base[start...].firstIndex(where: isSeparator)
+        ?? base.endIndex
+
+      if base[start..<end].isEmpty {
+        if omittingEmptySubsequences {
+          // Find the next subsequence of non-separators.
+          start =
+            base[end...].firstIndex(where: { !isSeparator($0) })
+            ?? base.endIndex
+          if start == base.endIndex {
+            // No non-separators left in the base collection. We're done.
+            return endIndex
+          }
+          end = base[start...].firstIndex(where: isSeparator) ?? base.endIndex
+        }
+      }
+    }
+
+    var updatedSplitCount = splitCount
+    if end < base.endIndex {
+      // This subsequence ends on a separator (and perhaps includes other
+      // separators, if we're omitting empty subsequences), so we've performed
+      // another split.
+      updatedSplitCount += 1
+    }
+
+    return Index(
+      baseRange: start..<end,
+      sequenceLength: sequenceLength + 1,
+      splitCount: updatedSplitCount
+    )
+  }
+
+  @inlinable
+  public var startIndex: Index {
+    _startIndex
+  }
+
+  @inlinable
+  public var endIndex: Index {
+    _endIndex
+  }
+
+  @inlinable
+  public func index(after i: Index) -> Index {
+    precondition(i != endIndex, "Can't advance past endIndex")
+
+    var subsequenceStart = i.baseRange.upperBound
+    if subsequenceStart < base.endIndex {
+      // If we're not already at the end of the base collection, the previous
+      // susequence ended with a separator. Start searching for the next
+      // subsequence at the following element.
+      subsequenceStart = base.index(after: i.baseRange.upperBound)
+    }
+
+    guard subsequenceStart != base.endIndex else {
+      if !omittingEmptySubsequences
+        && i.sequenceLength < i.splitCount + 1
+      {
+        // The base collection ended with a separator, so we need to emit one
+        // more empty subsequence. This one differs from `endIndex` in its
+        // `sequenceLength` (except in an extreme edge case!), which is the
+        // sole property tested for equality and comparison.
+        return Index(
+          baseRange: base.endIndex..<base.endIndex,
+          sequenceLength: i.sequenceLength + 1,
+          splitCount: i.splitCount
+        )
+      } else {
+        return endIndex
+      }
+    }
+
+    return indexForSubsequence(
+      atOrAfter: subsequenceStart,
+      sequenceLength: i.sequenceLength,
+      splitCount: i.splitCount
+    )
+  }
+
+  @inlinable
+  public subscript(position: Index) -> Base.SubSequence {
+    precondition(position != endIndex, "Can't subscript using endIndex")
+    return base[position.baseRange]
+  }
+}
+
+extension LazySplitCollection.Index: Hashable where Base.Index: Hashable {}
```

Comment:
I think the way we have it implemented now, the synthesized conformance would work, but generally when you implement custom `Equatable` conformance that only considers a subset of a type's members, the `Hashable` conformance should have the same customization. Can you implement `hash(into:)` here? (Interestingly, with that done, we also shouldn't need the conditional constraint anymore.)